CONTRIBUTING.md

```diff
@@ -0,0 +1,77 @@
+# Contributing guidelines
+
+Thank you for your interest in contributing to `dusk-blockchain`! In this file, you will find some guidelines to adhere to when opening issues, pull requests, and committing code, as well as writing code itself. Following these guidelines ensures a smooth review process, and increases likelihood of your code getting merged.
+
+## Coding conventions
+
+* Ensure to document all exported functions, methods, and structs, according to the [Golang documentation guidelines](https://blog.golang.org/godoc).
+* Avoid use of gotos as they lead to spaghetti code.
+* No else in the if statement, should be the default. Try to always limit the amount of indentation used, as it might make it harder to read code.
+* New code always needs to be tested. Pull requests will generally not be merged if there is no proper test coverage of newly added code.
+* If you fix a bug which was hard to find, create/add to the `gotchas.md` file in the package. This helps other developers catch the bug much quicker if it pops up again.
+* If you are adding new code, please include proper documentation outside of code comments where it makes sense. Ideally, for adding new features, it helps to at the very minimum add a `README.md` which explains purpose, inner workings, etc. In some cases, it may also be helpful to include UML diagrams which depict the new feature/component (sequence diagrams, component diagrams and workflows are often good choices).
+* Use the `context` package to manage goroutine lifetimes. For good examples, look at how the `Chain` and `Mempool` are managed using contexts.
+
+### Performance
+
+* Wherever possible, when allocating slices, try to allocate the entire slice in one go, and then assigning to the indices (or using `copy` where applicable), instead of creating an empty slice and then calling `append`. This prevents an unnecessary amount of dynamic memory allocations.
+* When creating a channel which sole purpose is to send signals, and not data, use a `chan struct{}`. The empty struct in Go takes up 0 bytes of space, and is thus the most efficient way of sending signals.
+* When clearing a map, use a `range` loop, and delete every key. This is more efficient than replacing the map with a new instance, as the GC does not have to clean this up later.
+
```

Comment:
I would add a `### Concurrency` section, explaining the philosophy of avoiding concurrent components unless absolutely necessary, and to explain when `channels`, `mutexes` or `atomic` are to be preferred (or link to some external resource for more in-depth explanations).